### ISSUE TITLE
Some dirt block remain after they have been dug

### ISSUE THREAD

**Issue by akien-mga** [OpenDungeons/OpenDungeons#51]
I experienced a small inconsistency in the digging algorithm and/or communication with server during a multiplayer game, a shown by this screenshot:

![opendungeons-bug-gold-leftovers](https://cloud.githubusercontent.com/assets/4701338/3865497/c0a55dee-1f98-11e4-9402-cfbf915d50e9.png)

I had marked the whole rectangular area (half claimed on the screenshot) to be dug, and the yellow digging creater took on the task meticulously.

For three of the blocks, I saw it digging through it, but the block remained unaffected while the creature continued with the next block. As you can see on the screenshot, the three blocks are still marked even when the area is being claimed, and they are claimed as walls (or pillars in this case).

**Comment by akien-mga** [comment]
Could be related, some other funny glitches (golden walls):

![odscreenshot_1](https://cloud.githubusercontent.com/assets/4701338/3865732/23a7572c-1fa7-11e4-9b68-df03f3635338.png)
![odscreenshot_2](https://cloud.githubusercontent.com/assets/4701338/3865734/2680e756-1fa7-11e4-825c-939aaebfad0f.png)

EDIT: Actually the above two screenshots were taken using my Intel GPU, so it might be more related to issue #41. The screenshot in the first comment was taken with my Nvidia GPU.

**Comment by Bertram25** [comment]
Good catch!

The golden walls are in fact the first dirt model loaded with a normal map:
https://github.com/OpenDungeons/OpenDungeons/blob/development/source/RenderManager.cpp#L1310
It's the same thing than when you see a shiny wizard, next to a not shiny one. And it is another issue.

But the ghost walls are indeed something to fix, even if I have no clue for now. :)

**Comment by Bertram25** [comment]
Note to self: To be tested while checking #59

**Comment by hwoarangmy** [comment]
Note to self : I've seen it in single player after #59. But no clue why. I've paid attention to see the consequences. It seems the bug is only visual as the creatures go through and you cannot dig it. I did not try to add a room on it but I guess it would have been possible. If someone sees it, it would be an interesting test (it would make us sure that the server considers the tile as being dirt).

**Comment by Bertram25** [comment]
I think I've found a way to reproduce. It's a bit tricky though:
- Wait for a worker to start dancing in front of a dirt wall tile. The worker is currently trying to turn it into a claimed wall.
- Mark it as it is still a dirt wall tile.
- The tile will turn into a marked claimed wall tile. (Everything is normal at the moment)
- The worker will then dig it right way usually.
- The marked claimed wall tile stays as is even if it should be dug.

The game handles the walls as ground afterwards.

**Comment by hwoarangmy** [comment]
Interesting. That would explain why it doesn't happen that much...

**Comment by akien-mga** [comment]
I will try your procedure to reproduce it.

I'm not sure this is the only case though, IIRC, the original issue (see the screenshot in the first comment) happened when I had marked the whole rectangular area for digging, and then let the workers do their work without human interaction.

As can be seen on the screenshot, the faulty tiles are either a marked dirt wall tile (through which creatures can go, so it's probably only on client side), or a marked claimed wall tile once a worker has claimed the (on server side) dirt floor tile which is under (and the marked claimed wall can also be walked through IIRC).

**Comment by akien-mga** [comment]
> The golden walls are in fact the first dirt model loaded with a normal map:
> https://github.com/OpenDungeons/OpenDungeons/blob/development/source/RenderManager.cpp#L1310
> It's the same thing than when you see a shiny wizard, next to a not shiny one. And it is another issue.

Do we have an open issue for this?

**Comment by hwoarangmy** [comment]
I gave a try but could not reproduce :(

> I'm not sure this is the only case though, IIRC, the original issue

I believe a did the same. But we can hope it will point out the same root.

**Comment by Bertram25** [comment]
I'll make more tries but it's not something easy to reproduce, unfortunately.

> The golden walls are in fact the first dirt model loaded with a normal map:
> https://github.com/OpenDungeons/OpenDungeons/blob/development/source/RenderManager.cpp#L1310
> It's the same thing than when you see a shiny wizard, next to a not shiny one. And it is another issue.

It's something I want to fix while on #60 in fact, as the catalog will permit saner appliance of normals on models.

**Comment by akien-mga** [comment]
> > The golden walls are in fact the first dirt model loaded with a normal map:
> > https://github.com/OpenDungeons/OpenDungeons/blob/development/source/RenderManager.cpp#L1310
> > It's the same thing than when you see a shiny wizard, next to a not shiny one. And it is another issue.
> 
> It's something I want to fix while on #60 in fact, as the catalog will permit saner appliance of normals on models.

I hadn't tried TestSingleplayer.level much (it's great!), but I did it now and noticed something interesting: the normal map issue is not present. The shiny dirt walls are easily reproducible in the small test map, but I can't reproduce them in the big one. Same thing for the creatures, I did not spot any gleaming creature in the big map :-)

Can you confirm this observation?

**Comment by akien-mga** [comment]
I managed to reproduce the ghost tiles issue in TestSingleplayer, while building @hwoarangmy's new pull request :-) This was a nice way to dramatically increase my latency (or decrease my computer's reactivity), so maybe the root of the problem is a latency issue in the communication between the server and the client? If some information is lost, we might end up with the client and the server having a different conception of the map, no?

**Comment by Bertram25** [comment]
> Can you confirm this observation?

I'm not confirming it, sorry. They are more hidden, but last time I tried they were here.

**Comment by hwoarangmy** [comment]
> If some information is lost, we might end up with the client and the server having a different conception of the map, no?

Yes, of course. But we use TCP sockets and we always expect packets sent to be received. If some of them are lost, it would be a really big problem...

**Comment by hwoarangmy** [comment]
As stated here http://forum.freegamedev.net/viewtopic.php?f=45&t=4765&p=58933#p58933
I could reproduce the bug by making the server send a tile fullness dug message to an already dug tile. Even if I don't know the true root of the bug, we can fix it by adding a sanity check before changing the tile fullness on the client side. I will add some asserts around the tilefullness change message, through, to try to catch the root if possible...

**Comment by Bertram25** [comment]
> we can fix it by adding a sanity check before changing the tile fullness on the client side.

+1 You found out the a way to reproduce. Congrats. :D 

> I will add some asserts around the tilefullness change message, through, to try to catch the root if possible...

+1 as well.

**Comment by hwoarangmy** [comment]
I've found out what was going on. It was not a problem of 2 messages being sent after all. In fact, the problem came from :
https://github.com/OpenDungeons/OpenDungeons/blob/development/source/Tile.cpp#L100

When called :
https://github.com/OpenDungeons/OpenDungeons/blob/development/source/Tile.cpp#L1146

If digRate is a little bit greater than the fullness, fullness - digRate < 1 and the tile is considered digged but no message is sent to the clients.
Patch will come soon :)

**Comment by akien-mga** [comment]
Awesome! Looking forward to testing your patch :-)

**Comment by hwoarangmy** [comment]
> Awesome! Looking forward to testing your patch :-)

Not that easy since bug was pretty random but the patch is there. I've seen there is a conflict but I'm gonna try to resolve it in my repository if possible :)

**Comment by hwoarangmy** [comment]
The conflict is resolved :)

**Comment by Bertram25** [comment]
Closed with 892a21aba62957f0c33980efdd68932f2f2ad8fb
